app/src/ui/lib/section-filter-list.tsx

```diff
@@ -0,0 +1,687 @@
+  readonly matches: IMatches
+}
+
+/**
+ * A row in the list. This is used internally after the user-provided groups are
+ * flattened.
+ */
+type IFilterListRow<T extends IFilterListItem> =
+  | IFlattenedGroup
+  | IFlattenedItem<T>
+
+interface ISectionFilterListProps<T extends IFilterListItem> {
+  /** A class name for the wrapping element. */
+  readonly className?: string
+
+  /** The height of the rows. */
+  readonly rowHeight: number
+
+  /** The ordered groups to display in the list. */
+  readonly groups: ReadonlyArray<IFilterListGroup<T>>
+
+  /** The selected item. */
+  readonly selectedItem: T | null
+
+  /** Called to render each visible item. */
+  readonly renderItem: (item: T, matches: IMatches) => JSX.Element | null
+
+  /** Called to render header for the group with the given identifier. */
+  readonly renderGroupHeader?: (identifier: string) => JSX.Element | null
+
+  /** Called to render content before/above the filter and list. */
+  readonly renderPreList?: () => JSX.Element | null
+
+  /**
+   * This function will be called when a pointer device is pressed and then
+   * released on a selectable row. Note that this follows the conventions
+   * of button elements such that pressing Enter or Space on a keyboard
+   * while focused on a particular row will also trigger this event. Consumers
+   * can differentiate between the two using the source parameter.
+   *
+   * Note that this event handler will not be called for keyboard events
+   * if `event.preventDefault()` was called in the onRowKeyDown event handler.
+   *
+   * Consumers of this event do _not_ have to call event.preventDefault,
+   * when this event is subscribed to the list will automatically call it.
+   */
+  readonly onItemClick?: (item: T, source: ClickSource) => void
+
+  /**
+   * This function will be called when the selection changes as a result of a
+   * user keyboard or mouse action (i.e. not when props change). This function
+   * will not be invoked when an already selected row is clicked on.
+   *
+   * @param selectedItem - The item that was just selected
+   * @param source       - The kind of user action that provoked the change,
+   *                       either a pointer device press, or a keyboard event
+   *                       (arrow up/down)
+   */
+  readonly onSelectionChanged?: (
+    selectedItem: T | null,
+    source: SelectionSource
+  ) => void
+
+  /**
+   * Called when a key down happens in the filter text input. Users have a
+   * chance to respond or cancel the default behavior by calling
+   * `preventDefault()`.
+   */
+  readonly onFilterKeyDown?: (
+    event: React.KeyboardEvent<HTMLInputElement>
+  ) => void
+
+  /** Called when the Enter key is pressed in field of type search */
+  readonly onEnterPressedWithoutFilteredItems?: (text: string) => void
+
+  /** Aria label for a specific group */
+  readonly groupAriaLabel?: (group: number) => string | undefined
```

Comment:
Nit: Since this is a function, it feels like it should be something like `getGroupAriaLabel`. I had a double take when I saw it's usage as a function (even tho it makes sense that groupAriaLabel would have to be a method like `renderGroupHeader`)

Alternatively, if the group objects simply had a `label` property, I think we wouldn't have a need for `renderGroupHeader` and this `getGroupAriaLabel` especially because usually the label is determined at the same component hierarchical level as the groups. But, I know that convention was already set in the original FilterList component, so may be out of scope for this.

Reply:
> Nit: Since this is a function, it feels like it should be something like `getGroupAriaLabel`.

👍 

> Alternatively, if the group objects simply had a `label` property…

True! I will take a look 👀